File: frontend-vite/src/utils/team_name_normalizer.ts
/**
 * Characters that survive NFD decomposition without becoming combining marks,
 * so they need an explicit expansion. This mirrors the backend normalizer.
 */
const extraMap: Record<string, string> = {
    "ß": "ss",
    "æ": "ae",
    "œ": "oe",
    "ø": "o",
    "đ": "d",
    "ð": "d",
    "þ": "th",
    "ł": "l"
};

/**
 * Search key for a team name: lowercased, diacritics stripped, whitespace
 * collapsed. Punctuation is deliberately preserved.
 *
 * This must stay behaviourally identical to normalizeTeamName() on the
 * backend, which decides which aliases are allowed to exist.
 */
export function normalizeForSearch(str: string): string {
    return str
        .toLowerCase()
        .replace(/[ßæœøđðþł]/g, char => extraMap[char] || char)
        .normalize("NFD")
        .replace(/\p{Mn}+/gu, "")
        // Match java.util.regex.Pattern's default ASCII \s class used by
        // the backend, rather than JavaScript's broader Unicode \s class.
        .replace(/[ \t\n\x0B\f\r]+/g, " ")
        .trim();
}

File: src/main/kotlin/eu/gaelicgames/referee/util/TeamNameNormalizer.kt
package eu.gaelicgames.referee.util

import java.text.Normalizer

/**
 * Characters that carry no combining mark and therefore survive NFD
 * decomposition, so they need an explicit expansion.
 * Mirrors the map in frontend-vite/src/utils/team_name_normalizer.ts.
 */
private val EXTRA_EXPANSIONS = mapOf(
    'ß' to "ss",
    'æ' to "ae",
    'œ' to "oe",
    'ø' to "o",
    'đ' to "d",
    'ð' to "d",
    'þ' to "th",
    'ł' to "l"
)

private val COMBINING_MARKS = Regex("\\p{Mn}+")
private val WHITESPACE_RUN = Regex("\\s+")

/**
 * Search/uniqueness key for a team name: lowercased, diacritics stripped,
 * whitespace collapsed. Punctuation is deliberately preserved, so
 * "Saint-Brieuc" and "Saint Brieuc" stay distinct.
 *
 * This is the only place team names may be normalized on the backend.
 * Changing these rules requires re-normalizing every TeamAliases row.
 */
fun normalizeTeamName(input: String): String {
    val expanded = buildString {
        for (char in input.lowercase()) {
            append(EXTRA_EXPANSIONS[char] ?: char)
        }
    }
    return Normalizer.normalize(expanded, Normalizer.Form.NFD)
        .replace(COMBINING_MARKS, "")
        .replace(WHITESPACE_RUN, " ")
        .trim()
}
